contracts/adapters/SwapAdapter.sol

```diff
@@ -0,0 +1,262 @@
+        _;
+    }
+
+    // Admin functions
+    function setTokenResourceID(address token, bytes32 resourceID) external onlyAdmin {
+        if (tokenToResourceID[token] == resourceID) revert AlreadySet();
+        tokenToResourceID[token] = resourceID;
+        emit TokenResourceIDSet(token, resourceID);
+    }
+
+    /**
+        @notice Function for depositing tokens, performing swap to ETH and bridging the ETH.
+        @param destinationDomainID  ID of chain deposit will be bridged to.
+        @param recipient Recipient of the deposit.
+        @param gas The amount of gas needed to successfully execute the call to recipient on the destination. Fee amount is
+                   directly affected by this value.
+        @param message Arbitrary encoded bytes array that will be passed as the third argument in the
+                       ISygmaMessageReceiver(recipient).handleSygmaMessage(_, _, message) call. If you intend to use the
+                       DefaultMessageReceiver, make sure to encode the message to comply with the
+                       DefaultMessageReceiver.handleSygmaMessage() message decoding implementation.
+        @param token Input token to be swapped.
+        @param tokenAmount Amount of tokens to be swapped.
+        @param amountOutMinimum Minimal amount of ETH to be accepted as a swap output.
+        @param pathTokens Addresses of the tokens for Uniswap swap. WETH address is used for ETH.
+        @param pathFees Fees for Uniswap pools.
+    */
+    function depositTokensToEth(
+        uint8 destinationDomainID,
+        address recipient,
+        uint256 gas, 
+        bytes calldata message,
+        address token,
+        uint256 tokenAmount,
+        uint256 amountOutMinimum,
+        address[] memory pathTokens,
+        uint24[] memory pathFees
+    ) external {
+        if (tokenToResourceID[token] == bytes32(0)) revert TokenInvalid();
+
+        // Swap all tokens to ETH (exact input)
+        IERC20(token).safeTransferFrom(msg.sender, address(this), tokenAmount);
+        IERC20(token).safeApprove(address(_swapRouter), tokenAmount);
+
+        uint256 amount;
+
+        {
+            bytes memory path = _verifyAndEncodePath(
+                pathTokens,
+                pathFees,
+                token,
+                _weth
+            );
+            IV3SwapRouter.ExactInputParams memory params = IV3SwapRouter.ExactInputParams({
+                path: path,
+                recipient: address(this),
+                amountIn: tokenAmount,
+                amountOutMinimum: amountOutMinimum
+            });
+
+            amount = _swapRouter.exactInput(params);
+        }
+
+        emit TokensSwapped(_weth, amount);
+        IWETH(_weth).withdraw(amount);
+
+        // Make Native Token deposit
+        _nativeTokenAdapter.depositToEVMWithMessage{value: amount}(
+            destinationDomainID,
+            recipient,
+            gas,
+            message
```

Comment:
I feel like we need to support both call with and without optional message. Maybe we need separate methods for that?

Reply:
Thanks for clarifying! Updated, made separate functions for that